src/torch/README.md

```diff
@@ -0,0 +1,123 @@
+
+*   **Automatic Differentiation (Autodiff):** A technique for numerically evaluating the derivative of a function specified by a computer program.  PyTorch uses a "reverse mode" autodiff system.
+
+*   **Reverse Mode Autodiff:**  In reverse mode, the gradients are computed by traversing the computational graph backwards, from the output (typically a scalar loss function) to the inputs.  This is highly efficient for functions with many inputs and a single output, which is common in machine learning.
+
+*   **Jacobian-Vector Product (JVP):**  Let `y = f(x)` be a vector-valued function, where `x` is an input vector and `y` is an output vector. The Jacobian matrix `J` contains all the first-order partial derivatives of `f`.  The JVP is the product `Jv`, where `v` is a vector.  It gives the directional derivative of `f` at `x` in the direction of `v`.  *Forward mode* autodiff computes JVPs.
+
+*   **Vector-Jacobian Product (VJP):** With `y = f(x)` as above, the VJP is the product `w^T J`, where `w` is a vector (often called the "upstream/puput gradient").  The VJP represents the gradient of a scalar-valued function that depends on `y` (e.g., a loss function), with respect to the input `x`.  *Reverse mode* autodiff computes VJPs. 
+
+*   **Hessian-Vector Product (HVP):** The Hessian matrix `H` contains all the second-order partial derivatives of a scalar-valued function `f(x)`. The HVP is the product `Hv`, where `v` is a vector.  It gives the second-order directional derivative of `f` at `x` in the direction of `v`. 
+
+## Wrapper Design
+
+The wrapper provides three main classes:
+
+1.  `SIRFTorchOperator`: Wraps a SIRF `Operator` (e.g., a projection operator). Computes the JVP forward and VJP in reverse.
+2.  `SIRFTorchObjectiveFunction`: Wraps a SIRF `ObjectiveFunction` for calculating its value forward, and gradient in reverse mode.
+3.  `SIRFTorchObjectiveFunctionGradient`: Wraps a SIRF `ObjectiveFunction` for calculating its gradient forward and computing the Hessian-vector product in reverse mode.
+
+These classes use custom `torch.autograd.Function` implementations (`_Operator`, `_ObjectiveFunction`, and `_ObjectiveFunctionGradient`) to define the forward and backward passes, handling the conversions between PyTorch tensors and SIRF objects.
+
+### `_Operator` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Applies the SIRF `Operator.forward()` method.  Let's represent the SIRF Operator as `A`.  So, this step computes `y = Ax`, where `x` is the input (SIRF object) and `y` is the output (SIRF object).
+    3.  Converts the result back to a PyTorch tensor.
+    4.  If the input tensor requires gradients, it saves relevant information (the output SIRF object and the operator) in the context (`ctx`) for use in the backward pass.
+
+*   **Backward Pass (Adjoint Operation - VJP):**
```

Comment:
it is only the adjoint (applied to a vector) in the linear case. I'd delete this. It is just VJP.

Reply:
I guess jacobian transpose/adjoint could work, but I will remove

Reply:
comment was resolved, but text removed yet.